File: ImasLiveDB/Domain/UseCases/EventGrouping.swift
import Foundation

/// イベント一覧の「年度グループ」。`groupEventsByYear` の出力単位。
struct YearGroup: Identifiable {
    var id: String { year }
    let year: String
    let events: [EventWithDate]
}

/// 時系列フィルタ + 年度グルーピング。
///
/// 本体は imas-core (Rust) の `domain/event_grouping.rs`。今後/開催済みの境界判定・
/// 部分日付 ("YYYY" 等) の桁合わせ・「年度不明」を末尾に置く規則もそちらに記載。
/// ここは firstDate の射影を 1 回の FFI 呼び出しに渡し、返った index 列を
/// `events` から引き直して表示用 `YearGroup` に組み立てるだけ (件数によらず 1 呼び出し)。
///
/// `todayKey` の既定値は注入しない: 呼び出し側が `JSTDay.today()` を渡す
/// (公演日は日本の開催日なので、端末ローカル日で切ると海外で 1 日ずれる)。
func groupEventsByYear(_ events: [EventWithDate], upcoming: Bool, todayKey: String) -> [YearGroup] {
    let groups = groupEventIndicesByYear(
        firstDates: events.map(\.firstDate),
        upcoming: upcoming,
        todayKey: todayKey
    )
    return groups.map { group in
        YearGroup(year: group.year, events: group.indices.map { events[Int($0)] })
    }
}

File: imas-core/src/inbound/event_grouping.rs
//! イベント年度グルーピングの FFI 面。ロジックは domain::event_grouping。
//!
//! エンティティ全体ではなく初回公演日の射影 (`first_dates`) を受け、
//! 「年ラベル + index 列」([`EventYearGroup`]) を返す (呼び出し側が自国の配列を
//! index で引き直す)。1 ユーザー操作 = 1 呼び出し。

use crate::domain::event_grouping::EventYearGroup;

/// 時系列フィルタ + 年度グルーピング。今後/開催済みの境界・並び順の規則は
/// domain 側のドキュメント参照。`today_key` には JST の今日 (`jst_today`) を渡す。
#[uniffi::export]
pub fn group_event_indices_by_year(
    first_dates: Vec<Option<String>>,
    upcoming: bool,
    today_key: String,
) -> Vec<EventYearGroup> {
    crate::domain::event_grouping::group_events_by_year(&first_dates, upcoming, &today_key)
}
